lib/Net/DRI/Protocol/EPP/Extensions/JP/JPEX.pm

```diff
@@ -0,0 +1,346 @@
+
+ $mes->command_body(\@d);
+
+ my @jpex;
+ Net::DRI::Exception::usererr_insufficient_parameters('suffix missing: jp or ojp') unless $rd->{'suffix'};
+ if ($rd->{'suffix'} && $rd->{'suffix'} eq 'jp') {
+  Net::DRI::Exception::usererr_insufficient_parameters('ryid and handle missing') unless ( $rd->{'ryid'} && $rd->{'handle'} );
+  push @jpex,['jpex:domain suffix="jp" transfer="domain"'];
+  push @jpex,['jpex:contact', ['jpex:handle', $rd->{'ryid'}], {'alloc'=>'registrant'}];
+  push @jpex,['jpex:contact', ['jpex:handle', $rd->{'handle'}], {'alloc'=>'public'}];
+ } elsif ($rd->{'suffix'} && $rd->{'suffix'} eq 'ojp') { # TODO: still need to confirm the following :(
+  push @jpex,['jpex:domain suffix="ojp" transfer="agent"'];
+ }
+
+ my $eid=build_command_extension($mes,$epp,'jpex:transfer');
+ $mes->command_extension($eid,[@jpex]);
+
+ return;
+}
+
+# README: by the documentation is seems that this is only acceptable 10 days before the expiry date
+# added trade (as per documentation) but such is not possible todo. in order to change contact registrant object details
+# this need to be done via a <contact:update> - not possible to use <domain:update> to set a new registrant as well!
+sub trade_request
+{
+ my ($epp,$domain,$rd)=@_;
+ my $mes=$epp->message();
+ my @d=Net::DRI::Protocol::EPP::Util::domain_build_command($mes,['transfer',{'op'=>'request'}],$domain);
+ push @d,['domain:period', 0, {'unit'=>'y'}]; # always enforce 0
+
+ $mes->command_body(\@d);
+
+ my @jpex;
+ Net::DRI::Exception::usererr_insufficient_parameters('ryid and handle missing') unless ( $rd->{'ryid'} && $rd->{'handle'} );
+ push @jpex,['jpex:domain suffix="jp" transfer="registrant"'];
+ push @jpex,['jpex:contact', ['jpex:ryid', $rd->{'ryid'}], {'alloc'=>'registrant'}];
+ push @jpex,['jpex:contact', ['jpex:handle', $rd->{'handle'}], {'alloc'=>'public'}];
+
+ my $eid=build_command_extension($mes,$epp,'jpex:transfer');
+ $mes->command_extension($eid,[@jpex]);
+
+ return;
+}
+
+
+
+sub _organization
+{
+ my ($rd)=@_;
+ my @organization;
+ if ($rd && (ref($rd) eq 'HASH')) {
+  foreach my $o (qw/name nameKana nameEn postalCode address addressEn type typeEn registrationDate registrationAddress replyMail request/) {
+   push @organization,['jpex:'.$o, $rd->{$o}] if $rd->{$o};
+   # README: i think that `request` is a attribute but not sure/couldn't test in their OT&E!
+  }
+ }
+
+ return @organization;
+}
+
+
+sub _representative
+{
+ my ($rd)=@_;
+ my @representative;
+ if ($rd && (ref($rd) eq 'HASH')) {
+  foreach my $r (qw/lastName firstName lastNameEn firstNameEn title/) {
+   push @representative,['jpex:'.$r, $rd->{$r}] if $rd->{$r};
+  }
+ }
+
+ return @representative;
+}
+
+
+sub _condition
+{
+ my ($rd)=@_;
+ my @condition;
+ if ($rd && (ref($rd) eq 'HASH')) {
+  foreach my $c (qw/no_other_ojp_domain_is_registered_under_the_same_corporation the_applicant_must_a_registered_corporation/) {
```

Comment:
whats going on here? :) Is my years of not actually coding failing me, or is this only one _long_ variable in the foreach? Not the end of the world...

Reply:
Intended to build what's listed in their XML shema for: ojpConditionType element (as organizationType and representativeType). Like this the function is already implemented in case they add/remove anything in the future.